test/core/storage/trie/polkadot_trie/polkadot_trie_test.cpp

```diff
@@ -333,6 +333,147 @@ TEST_F(TrieTest, ClearPrefix) {
   ASSERT_TRUE(trie->empty());
 }
 
+struct DeleteData {
+  std::vector<Buffer> data;
+  Buffer key;
+  size_t size;
+};
+
+class DeleteTest : public testing::Test,
+                   public ::testing::WithParamInterface<DeleteData> {
+ public:
+  DeleteTest() {}
+
+  void SetUp() override {
+    trie = std::make_unique<PolkadotTrieImpl>();
+  }
+
+  std::unique_ptr<PolkadotTrieImpl> trie;
+};
+
+size_t size(const PolkadotTrie::NodePtr &node) {
+  size_t count = 0;
+  if (node) {
+    if (node->isBranch()) {
+      auto branch =
+          std::dynamic_pointer_cast<kagome::storage::trie::BranchNode>(node);
+      for (const auto &child : branch->children) {
+        count += size(child);
+      }
+    }
+    ++count;
+  }
+  return count;
+}
+
+TEST_P(DeleteTest, DeleteData) {
+  for (auto &entry : GetParam().data) {
+    EXPECT_OUTCOME_TRUE_1(trie->put(entry, "123"_buf));
+  }
+  EXPECT_OUTCOME_TRUE_1(trie->remove(GetParam().key));
+  ASSERT_EQ(size(trie->getRoot()), GetParam().size);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    DeleteSuite,
+    DeleteTest,
+    testing::ValuesIn(
+        {DeleteData{{}, "bar"_buf, 0},
+         DeleteData{{"bar"_buf, "foo"_buf}, "bar"_buf, 1},
+         DeleteData{{""_buf, "bar"_buf, "foo"_buf}, "bar"_buf, 2},
+         DeleteData{{"bar"_buf, "foa"_buf, "fob"_buf}, "bar"_buf, 3},
+         DeleteData{{"612355"_hex2buf, "6124"_hex2buf}, "6123"_hex2buf, 3},
+         DeleteData{{"b"_buf, "ba"_buf, "bb"_buf}, "b"_buf, 3},
+         DeleteData{{"a"_buf, "b"_buf, "z"_buf}, "z"_buf, 3}}));
+
+struct ClearPrefixData {
+  std::vector<Buffer> data;
+  Buffer prefix;
+  boost::optional<uint64_t> limit;
+  std::vector<Buffer> res;
+  std::tuple<bool, uint32_t> ret;
+  size_t size;
+};
+
+class ClearPrefixTest : public testing::Test,
+                        public ::testing::WithParamInterface<ClearPrefixData> {
+ public:
+  ClearPrefixTest() {}
+
+  void SetUp() override {
+    trie = std::make_unique<PolkadotTrieImpl>();
+  }
+
+  std::unique_ptr<PolkadotTrieImpl> trie;
+};
+
+TEST_P(ClearPrefixTest, ManyCases) {
+  for (const auto &entry : GetParam().data) {
+    EXPECT_OUTCOME_TRUE_1(trie->put(entry, "123"_buf));
+  }
+  EXPECT_OUTCOME_TRUE_2(
+      ret,
+      trie->clearPrefix(
+          GetParam().prefix, GetParam().limit, [](const auto &, auto &&) {
+            return outcome::success();
+          }));
+  EXPECT_EQ(ret, GetParam().ret);
+  for (const auto &entry : GetParam().res) {
+    ASSERT_TRUE(trie->contains(entry));
+  }
+  ASSERT_EQ(size(trie->getRoot()), GetParam().size);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    ClearPrefixSuite,
+    ClearPrefixTest,
+    testing::ValuesIn(
+        /* empty tree */
```

Comment:
Sad, but not a given-when-then notation. Please provide them if possible. Pretty sure it is.
Does an "empty tree" relate to an initial state or that should be the result?